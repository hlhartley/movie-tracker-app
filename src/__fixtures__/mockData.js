export const mockMovieData = [
    {   
        vote_count: 3119,
        id: 297802,
        video: false,
        vote_average: 6.9,
        title: "Aquaman",
        popularity: 483.065,
        poster_path: "/5Kg76ldv7VxeX9YlcQXiowHgdX6.jpg",
        original_language: "en",
        original_title: "Aquaman",
        backdrop_path: "/5A2bMlLfJrAfX9bqAibOL2gCruF.jpg",
        adult: false,
        overview: "Once home to the most advanced civilization on Earth, the city of Atlantis is now an underwater kingdom ruled by the power-hungry King Orm. With a vast army at his disposal, Orm plans to conquer the remaining oceanic people -- and then the surface world. Standing in his way is Aquaman, Orm's half-human, half-Atlantean brother and true heir to the throne. With help from royal counselor Vulko, Aquaman must retrieve the legendary Trident of Atlan and embrace his destiny as protector of the deep.",
        release_date: "2018-12-07",
    },
    {
        vote_count: 1020,
        id: 424783,
        video: false,
        vote_average: 6.5,
        title: "Bumblebee",
        popularity: 263.575,
        poster_path: "/fw02ONlDhrYjTSZV8XO6hhU3ds3.jpg",
        original_language: "en",
        original_title: "Bumblebee",
        backdrop_path: "/hMANgfPHR1tRObNp2oPiOi9mMlz.jpg",
        adult: false,
        overview: "On the run in the year 1987, Bumblebee finds refuge in a junkyard in a small Californian beach town. Charlie, on the cusp of turning 18 and trying to find her place in the world, discovers Bumblebee, battle-scarred and broken. When Charlie revives him, she quickly learns this is no ordinary yellow VW bug.",
        release_date: "2018-12-15",
    },
    { 
        vote_count: 829,
        id: 480530,
        video: false,
        vote_average: 6.6,
        title: "Creed II",
        popularity: 191.233,
        poster_path: "/v3QyboWRoA4O9RbcsqH8tJMe8EB.jpg",
        original_language: "en",
        original_title: "Creed II",
        backdrop_path: "/8yqLPNwNCtpOPc3XkOlkSMnghzw.jpg",
        adult: false,
        overview: "Follows Adonis Creed's life inside and outside of the ring as he deals with new found fame, issues with his family, and his continuing quest to become a champion.",
        release_date: "2018-11-21",
    }
]