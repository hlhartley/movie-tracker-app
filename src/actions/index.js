export const addMovies = (movies) => ({
  type: 'ADD_MOVIES',
  movies
});

export const loginUser = (id, name) => ({
  type: 'LOGIN_USER',
  id,
  name
});

export const updateError = (message) => ({
  type: 'UPDATE_ERROR',
  message
});

export const logoutUser = () => ({
  type: 'LOGOUT_USER'
});

export const showPopup = (bool) => ({
  type: 'SHOW_POPUP',
  showPopup: bool
});

export const toggleFavorite = (id) => ({
  type: 'TOGGLE_FAVORITE',
  id
});
