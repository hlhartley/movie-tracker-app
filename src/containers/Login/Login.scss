.form-container {
    display: flex;
    flex-flow: column;
    justify-content: center;
    align-items: center;
    line-height: 55px;
    font-size: 30px;
    font-weight: 700;
    height: 50vh;
    background-color: $brown;
    margin-top: 35px;
    width: 80%;
    color: $silver;
    border-radius: 5px;
}

.login-btn {
    margin-top: 20px;
    border-radius: 10px;
    width: 22vw;
    border: 1px solid black;
    background-color: $silver;
    font-weight: 700;
    padding: 5px;
}

input {
    width: 40vw;
    text-align: center;
    margin: 10px;
    margin-left: 20px;
}

.email-container,
.password-container {
    display: flex;
    justify-content: space-around;
}

.fa-user-circle {
    font-size: 60px;
    margin: 25px;
    color: $gold;
}