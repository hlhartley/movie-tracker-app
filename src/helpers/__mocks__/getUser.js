const mockUser = {
    data: {
        id: 2,
        name: 'John'
    }
}

export const getUser = jest.fn().mockImplementation(() => Promise.resolve(mockUser))