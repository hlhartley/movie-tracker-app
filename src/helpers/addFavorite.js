export const addFavorite = async (movie_id, user_id, title, poster_path, release_date, vote_average, overview) => {
  const response = await fetch('http://localhost:3000/api/users/favorites/new', {
    method: 'POST',
    body: JSON.stringify({
      movie_id, 
      user_id, 
      title, 
      poster_path, 
      release_date, 
      vote_average, 
      overview
    }),
    headers: {
      'Content-Type': 'application/json'
    }
  });

  if (response.status >= 300) {
    throw Error(`Error saving to favorites: ${response.statusText}`);
  } else {
    const result = await response.json();
    return result;
  }
}