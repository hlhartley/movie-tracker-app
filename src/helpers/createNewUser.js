export const createNewUser = async (user) => {

  const response = await fetch('http://localhost:3000/api/users/new', {
    method: 'POST',
    body: JSON.stringify(user),
    headers: {
      'Content-Type': 'application/json'
    }
  });

  if (response.status >= 300) {
    throw Error(`Email has already been used: ${response.statusText}`);
  } else {
    const result = await response.json();
    return result;
  }

}