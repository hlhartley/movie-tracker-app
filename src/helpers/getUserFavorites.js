export const getUserFavorites = async (user_id) => {
    const response = await fetch(`http://localhost:3000/api/users/${user_id}/favorites`)
    if(response.status >= 300) {
        throw Error(`Error getting favorites: ${response.statusText}`)
    } else {
        const result = response.json()
        return result
    }
}