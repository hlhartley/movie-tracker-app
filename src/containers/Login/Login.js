import React, { Component } from 'react';
import { getUser } from '../../helpers/requests';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { loginUser } from '../../actions';

export class Login extends Component {
    constructor() {
        super();
        this.state = {
            email: '',
            password: '',
            error: ''
        }
    }

    handleChange = (e) => {
        let { name, value } = e.target
        this.setState({ [name]: value })
    }

    handleSubmit = async (e) => {
        e.preventDefault()

        try {
            const result = await getUser(this.state)
            this.props.loginUser(result.data.id, result.data.name)
        } catch(error) {
            this.setState({ error: error.message })
        }
        
    }

    render() {
        if(!this.props.currentUser) {
            return(
                <form onSubmit={this.handleSubmit}>
                    <label htmlFor='email'>E-mail</label>
                    <input name='email' value={this.state.email} id='email' onChange={this.handleChange}/>
                    <label htmlFor='password'>Password</label>
                    <input name='password' value={this.state.password} id='password' onChange={this.handleChange}/>
                    <button>Login</button>
                </form>
            )    
        } else {
            return(
                <Redirect to='/'/>
            )
        }
    }
}

export const mapStateToProps = (state) => ({
    currentUser: state.currentUser
});

export const mapDispatchToProps = (dispatch) => ({
    loginUser: (id, name) => dispatch(loginUser(id, name))
});

export default connect(mapStateToProps, mapDispatchToProps)(Login);