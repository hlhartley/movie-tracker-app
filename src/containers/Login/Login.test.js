import { Login, mapStateToProps, mapDispatchToProps } from './Login';
import { shallow } from 'enzyme';
import { updateError, loginUser } from '../../actions';
import React from 'react';
import { getUser } from '../../helpers/getUser';
import { mockMovies } from '../../__fixtures__/mockData';

jest.mock('../../helpers/getUser.js')

describe('Login', () => {
  
  describe('Login component', () => {
    let wrapper;
    const mockLoginUser = jest.fn();
    const mockUpdateError = jest.fn()
    
    beforeEach(() => {
      let props = {
        currentUser: null, 
        errorStatus: '',
      }


      wrapper = shallow(<Login {...props} loginUser={mockLoginUser} updateError={mockUpdateError}/>)
    })

    it.skip('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    });

    it('should have a default state', () => {
      const expected = {
        email: '',
        password: ''
      }

      expect(wrapper.state()).toEqual(expected);
    });

    it('should handleChange from the email input', () => {
      const mockEvent = {
        target: {
          name: 'email',
          value: 'Lee@gmail.com'
        }
      }
      const expected = 'Lee@gmail.com'
      wrapper.find('#email').simulate('change', mockEvent)
      expect(wrapper.state('email')).toEqual(expected)
    });

    it('should handleChange from the password input', () => {
      const mockEvent = {
        target: {
          name: 'password',
          value: 'abc'
        }
      }
      const expected = 'abc'
      wrapper.find('#password').simulate('change', mockEvent)
      expect(wrapper.state('password')).toEqual(expected)
    });

    it('should handleSubmit, clearing any previous error messages', () => {
      const mockEvent = {
        preventDefault: jest.fn()
      }
      wrapper.find('form').simulate('submit', mockEvent)
      expect(mockUpdateError.mock.calls.length).toBe(1)
      expect(mockUpdateError).toHaveBeenCalledWith('')
    });

    it('should login user with correct params if everything is ok with getUser', async () => {
      const mockEvent = {
        preventDefault: jest.fn()
      }

      const expectedId = 2
      const expectedName = 'John'
    
      wrapper.find('form').simulate('submit', mockEvent)
      await expect(mockLoginUser).toHaveBeenCalledWith(expectedId, expectedName)
    });
  });

  describe('mapStateToProps', () => {
    it('should return an object with an error status and current user', () => {
      const mockState = {
        movies: mockMovies,
        currentUser: null,
        errorStatus: ''
      }
      const expected = {
        currentUser: null,
        errorStatus: '' 
      }

      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapStateToDispatch', () => {
    it('should call dispatch with a loginUser action when loginUser is called', () => {
      const mockDispatch = jest.fn();
      const id =1;
      const name = 'John';
      const actionToDispatch = loginUser(id, name);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.loginUser(id, name);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch with an updateError action when updateError is called', () => {
      const mockDispatch = jest.fn();
      const message = 'An error happened';
      const actionToDispatch = updateError(message);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.updateError(message);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
})