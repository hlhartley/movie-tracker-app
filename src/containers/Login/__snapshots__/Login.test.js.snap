// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Login component should match the snapshot 1`] = `
<form
  className="form-container"
  onSubmit={[Function]}
>
  <i
    className="fas fa-user-circle"
  />
  <div
    className="email-container"
  >
    <label
      htmlFor="email"
    >
      <i
        className="far fa-envelope"
      />
    </label>
    <input
      id="email"
      name="email"
      onChange={[Function]}
      placeholder="E-mail address"
      value=""
    />
  </div>
  <div
    className="password-container"
  >
    <label
      htmlFor="password"
    >
      <i
        className="fas fa-lock"
      />
    </label>
    <input
      id="password"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      value=""
    />
  </div>
  <button
    className="login-btn"
  >
    LOGIN
  </button>
</form>
`;
